model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java

```diff
@@ -1312,7 +1312,7 @@ public boolean removeClientScope(RealmModel realm, String id) {
 
     @Override
     public void removeClientScopes(RealmModel realm) {
-        realm.getClientScopesStream().map(ClientScopeModel::getId).forEach(id -> removeClientScope(realm, id));
+        getRealmDelegate().getClientScopesStream(realm).map(ClientScopeModel::getId).forEach(id -> removeClientScope(realm, id));
```

Comment:
Otherwise the `ClientModelTest#testClientScopes` is broken with `jpa+infinispan` profile. This call `session.clientScopes().removeClientScopes(realm);` doesn't remove client scopes.

Reply:
If I understand correctly the difference is that after your change `clientScopes` are loaded from the store instead of cache is it correct? If yes, why getting the scopes from the cache doesn't succeed? Should we fix the reason instead of loading it directly from the store?

Reply:
So problem with `realm.getClientScopesStream()` is that it looks into the cache which is empty, so nothing is removed eventually. So I dug into why the cache is empty and I see the client scopes are populated into `CachedRealm` only when `RealmCacheSession#getRealm` is called for the first time [1]. The method is called for the first time from `KeycloakModelTest#withRealm` when there are no client scopes created yet. Is there a flaw in model tests?

Second question is whether `removeClientScopes` should look only into cache and delete only what is cached. This might not work in cases when something is gone from the cache but still present in a storage. 

[1] https://github.com/keycloak/keycloak/blob/eae29422e4980dd6b3b725aa96dff12be873da01/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java#L414

Reply:
I would say the test is correct. Also, it seems the line you identified is indeed wrong, however, I am not sure the fix is the right one.

The way I understand it is that we are caching ids of all clientsScopes for each realm. When changing some client scope (addition/removal) we [invalidate realm](https://github.com/keycloak/keycloak/blob/main/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java#L1286-L1296). The problem of the method you identified is, that it is not checking invalidation and therefore in the test case you added to model tests it fails.

The way it is fixed here is that you removed looking into the cache at all, and we are just looking into the store directly.

I would say this is not necessary, we can  use the data in the cache if they are not stale, therefore  we need to use a method that is checking for invalidations, I think something like this should be sufficient:
```suggestion
        getClientScopesStream(realm).map(ClientScopeModel::getId).forEach(id -> removeClientScope(realm, id));
```

@vramik Could you please double-check?

Reply:
Thank you for the explanation. That makes sense.

Reply:
@mhajas I checked it and I believe your suggestion is correct, thank you!